### ISSUE TITLE
Seismogram and TimeSeries api in relation to Core equivalents

### ISSUE THREAD

**Issue by pavlis** [wangyinz/mspass#92]
The C++ api in MsPASS use a set of Core data objects that differ from their children (TimeSeries and Seismogram) only with the addition of ProcessingHistory to the non Core versions.  That is a useful design for two reasons:
1.  It makes all the code that is "Core" completely portable outside MsPASS so others might be able to utilize the code independent of MsPASS.   
2. It provides a clear message to anyone educated to this design that a function using a Core data object will not need or alter ProcessingHistory data.  

I'm creating this issues page to document how we might want to clean up or extend the current api to clarify this detail and insulate python users for the distinction.   i.e. I think we decided that our design should be such that python users don't need to even know about the existence of Core data objects.  They will only interact with Seismogram and TimeSeries objects.  

I just completed what I though was step one.  That is, I regularized the constructors for TimeSeries and Seismogram so that there are pretty much parallel constructors to Core equivalents.  That means I defined these for TimeSeries:
```
/*! Bare bones constructor allocates space and little else.

  Sometimes it is helpful to construct a skeleton that can be fleshed out
  manually.   This constructor allocates an nsamples vector and
  initializes it to all zeros.   The data are marked dead
  because the assumption is the caller will fill out commonly needed basic
  Metadata, load some kind of sample data into the data vector, and then
  call the set_live method when that process is completed.   That kind of
  manipulation is most common in preparing simulation or test data where
  the common tags on real data do not exist and need to be defined manually
  for the simulatioon or test.   The history section is initialized with
  the default constructor, which currently means it is empty.   If a simulation
  or test requires a history origin the user must load it manaually.

  \param nsamples is the number of samples needed for storing the data vector.
    */
  TimeSeries(const size_t nsamples);
  /*! Partially construct from components.

      There are times one wants to use the Metadata area as a template to
      flesh out a CoreTimeSeries as what might be called skin and bones:  skin is
      Metadata and bones as BasicTimeSeries data.   This constructor initializes
      those two base classes but does not fully a valid data vector.  It only
      attempts to fetch the number of points expected for the data vector using
      the npts metadata (integer) key (i.e. it sets npts to md.get_int("npts")).
      It then creates the data vector of that length and initialzies it to all zeros.

      This constructor is largely a wrapper for the CoreTimeSeries version
      with the same signature but it also initalizes the ProcessingHistory
      to null (calling the default constructor)*/
  TimeSeries(const BasicTimeSeries& bts,const Metadata& md);
```
and this one for Seismogram:
```
  /*! Bare bones constructor allocates space and little else.

  Sometimes it is helpful to construct a skeleton that can be fleshed out
  manually.   This constructor allocates a 3xnsamples array and sets the
  matrix to all zeros.  It also sets the orientation information to cardinal
  and defines the data in relative time units.  The data are marked dead
  because the assumption is the caller will fill out commonly needed basic
  Metadata, load some kind of sample data into the data matrix, and then
  call the set_live method when that process is completed.   That kind of
  manipulation is most common in preparing simulation or test data where
  the common tags on real data do not exist and need to be defined manually
  for the simulatioon or test.   The history section is initialized with
  the default constructor, which currently means it is empty.   If a simulation
  or test requires a history origin the user must load it manaually.

  \param nsamples is the number of 3c vector samples to iniitalize the
    object with (creates a 3xnsamples matrix).
    */
  Seismogram(const size_t nsamples);
```

Do we need others?   For CoreSeismogram we have this constructor not currently defined for Seismogram:
```
/*! \brief Construct from Metadata definition that includes data path.
 *
 A Metadata object is sufficiently general that it can contain enough
 information to contruct an object from attributes contained in it.
 This constuctor uses that approach, with the actual loading of data
 being an option (on by default).   In mspass this is constructor is
 used to load data with Metadata constructed from MongoDB and then
 using the path created from two parameters (dir and dfile used as
 in css3.0 wfdisc) to read data.   The API is general but the
 implementation in mspass is very rigid.   It blindly assumes the
 data being read are binary doubles in the right byte order and
 ordered in the native order for dmatrix (Fortran order).  i.e.
 the constuctor does a raw fread of ns*3 doubles into the internal
 array used in the dmatrix implementation.

 \param md is the Metadata used for the construction.  It MUST contain
 all of the following or it will fail:  delta, starttime,npts,U11,U21,
 U31,U21,U22,U23,U31,U32,U33,dir,dfile, and foff.

 \param load_data if true (default) a file name is constructed from
 dir+"/"+dfile, the file is openned, fseek is called to foff,
 data are read with fread, and the file is closed.  If false a dmatrix
 for u is still created of size 3xns, but the matrix is only initialized
 to all zeros.

 \exception  Will throw a MsPASSError if required metadata are missing.
 */
 CoreSeismogram(const mspass::utility::Metadata& md,const bool load_data=true);
```
I think that is correct as that constructor is a bit of an oddity anyway.  What I wonder more is if we need this:
```
Seismogram(const BasicTimeSeries&, const Metadata&);
```
to have a parallel api for a TimeSeries object with the same signature.  Perhaps Jinxin made a wrapper that does this in python, but it would seem sensible to have this parallel constructor with similar behavior available in python.

I'll wait to commit the first set of changes until I hear from you guys.   That last point about a new constructor for Seismogram would be very quick to implement.

**Comment by pavlis** [comment]
Forgot to say above that when we finalize api changes I'll plan to clean up the pybind11 code to not make CoreSeismogram or CoreTimeSeries even visible to python.

**Comment by wangyinz** [comment]
Looks good. I don't have a strong feeling about the last constructor, so I'll let @JiaoMaWHU decide whether he thinks it useful from a user's stand point. 

btw, I have mentioned it in a totally irrelevant issue, but I think we need to fix the problem where both `Metadata` and `ProcessingHistory` defined the `clear` method, so currently when a clear is called, it is actually the `Metadata` one. We need to at least rename one of them, or I think we might even turn `ProcessingHistory` into a member of the data object. What do you think?

**Comment by pavlis** [comment]
I will give the two clear methods different names in the pybind11 code.   That is cleaner and an appropriate use of the binding code.   The issue wouldn't appear in C++ because of syntax:  you have `ProcessingHistory::clear` and `Metadata::clear` that are unambiguous.

**Comment by pavlis** [comment]
I set the bindings in pybind11 for `ProcessingHistory::clear()` to use the more verbose and descriptive python name `clear_history()`.   That was trivial.

I could use a second opinion on `Metadata::clear`.   It has a different behavior.  The history version could have been called "clear_all" because it does that; completely flushes this history data.   A better name for the Metadata version would be "clear_one" because it takes a key and removes deletes that entry.   

Do you think we should have the wrapper rename `Metadata::clear` to be *clear* or *clear_one*?

**Comment by pavlis** [comment]
A final issue that comes up here is strongly you think we need to completely separate Core data objects from the working ones (i.e. TimeSeries and Seismogram) in the python interface?   

Reasons we might want to put Core objects only on the C side are:
1.  Simplicity - user's don' t have to worry about this detail.
2. Consistency - everything looks and behaves the same (well never really true, but there could be more uniformity.)
3. Potentially less error prone from reduced complexity in the api.

On the negative side:
1.  This would impose a pretty rigid rule that would likely be quickly violated if, as we hope, this package catches on and other people contribute.
2. It will be harder to adapt existing algorithms in C/C++ to the framework as they need to follow those rules. In particular,  an C++ algorithm would need to handle exceptions internally and post elog messages directly rather than depend on python wrappers.  Worse, this is an invitation to a different kind of inconsistency - either the C++ code or the python code could handle exceptions and it will be difficult to document such behavior.  
3. Experimental code, in particular, would need to be more conscious of the History and elog features.   Not sure many users will want that extra baggage when prototyping. 

My recommendation is we not be dogmatic about this at all but continue to support CoreSeismogram and CoreTimeSeries objects with these guidelines for use:
1.  Use the Core objects in a C++ implementation if the code only needs to handle errors that can be posted as ErrorSeverity.Invalid or worse.
2. C++ code that needs to post information log messages (ErrorSeverity.Complaint or lower) should not use the Core versions but operate on TimeSeries or Seismogram objects directly.
3.  More generally, any C++ code that doesn't need to to write to elog or call ProcessingHistory methods should use Core data objects.  The python interpreter, I think, will properly downcast Seismogram or TimeSeries to their core equivalents and not require any special wrappers.
4. To make it clear Core data objects should be special we should have pybind11 tag them as `_CoreTimeSeries` and `_CoreSeismogram`.  The underscore used this ways is a standard hint in python they should be used only be experts. 
5. Normal procedure for any algorithm written in C/C++ should be to use a python wrapper function to handle the ProcessingHistory and error log (actually remember elog is a member of ProcessingHistory at present) features.  All are lightweight and more appropriate in python.   
6. If an algorithm has methods for handling Seismogram or TimeSeries objects directly it must be document if a python wrapper is require or not.  How to do that is not at all obvious as it will not work with the automated doxygen and sphyx procedures.   We may need a deviates page in the documentation or something like that.   The basic rule probably should be that if a function/method has a Seismogram or TimeSeries object as an output it need to be responsible for ProcessingHistory and a wrapper should not be needed.  An issue, however, may be that such a thing may still need a wrapper to support the parallel API.  I'll leave that as a question to you guys?

Other rules?   

If we can finalize these we should put them in a first draft of an *Algorithm Development* guide.

**Comment by wangyinz** [comment]
About the `clear` method. I realized that I was wrong claiming that the name collides, it is actually only a confusion rather than a problem because the two methods are actually different on the python side `clear()` v.s. `clear(key)`. But still, we definitely want to rename them. I think `clear_history()` works perfect, and for the Metadata one, I think something like a `clear_key` makes more sense. Actually, now thinking more about it, I really think the word "clear" is not accurately describing the action here because I think it somehow implies removing a large chunk. Similar function in the map container of STL uses the word "erase", and I think maybe we should just follow that. Python's `dict` don't have such a method, and people usually call the `del` for that purpose. I have implemented that through the `__delitem__` magic method in Metadata's pybind11 code, so a more pythonic name may be "delete_key".

For the Core objects issue, I think the rules you listed were pretty good. One thought I have about the complexity you listed in item 6 is that maybe we should consider separating the elog from the ProcessingHistory. The problem really came from the need of C++ code to handle elog while it is now bundled with ProcessingHistory, which by design should always be handled on the Python side. Therefore, I guess we can make elog a member of Core objects, and this way, it does not matter whether an algorithm implemented in C++ or in Python posts something to the log. Then, we can simplify the rules such that  ProcessingHistory should always be handled on the Python side.

**Comment by pavlis** [comment]
Good points.  I will:
1.  Make clear in ProcessingHistory map to clear_history in python but leave the c code alone.
2.  Similarly I will map `Metadata::clear(const string key)`  to an erase(str) method in the python wrappers.  It meshes well with stl is, in fact, why I think I used clear for the method to avoid needing a namespace resolution on erase in the C++ code.   In both cases I think it best to leave the C++ code alone as it is too easy to create an unanticipated side effect (i.e. I'm being lazy and don't want to search C++ algorithm code for calls to a clear method.  
3. In contrast, I think i will "bite the bullet" and make the changes you suggest moving elog outside of ProcessingHistory.   I think it needs to continue to be a member  as opposed to a base class because an error log unambiguously satisifed the "has a" versus "is a" rule for the contents.   Purists, by the way, might argue Metadata also is more accurate with a "has a" definition, but our use is properly "is a" if you define the data as the sample data plus the parametric data required to work with it (a header).  elog is definitely different as it is optional.   
4. I think the question of if elog should be a member of CoreTimeSeries and CoreSeismogram versus TimeSeries and Seismogram is less clear.   The rule I adopted up till now is that Core objects should be functional and behave consistently outside of MsPASS.   That follows standard object design to lay out the design to maximize reuse.   i.e. one should try to design the class hierarchy to allow the inheritance graph to be pruned at points that have a well defined reuse point.  I think it more appropriate to put elog in TimeSeries and Seismogram because it involves a new concept not required by all Core objects; error logging.  In all current code I have followed a critical rule about error handling I may not have articulated well above:  If we are working with Core data, error conditions should cause the code to throw an exception (or use the class return an error code - needed sometimes for backward compatibility)  BUT if the data being ,manipulated are Seismogram or TimeSeries data the code should (a) post errors or messages to elog and (b) if the error is Invalid or above mark the data dead before returning.   Both behaviors need to be defined in documentation and that is where they belong.   So, I need to be convinced of the merits of putting elog into the Core objects, BUT I am certain elog should be a member of Seismogram and TimeSeries (as the variable elog) for reasons I'll expand on next.
5. Putting elog as an attribute of TimeSeries and Seismogram objects is advised for a different reason.  Putting it in ProcessingHistory reduced the code required, but mixed two concepts in the current ProcessingHistory, which is an OOP design error.   The code base is more extensible and maintainable if we fission elog from ProcessingHistory.   

I will make the pybind11 changes immediately.  Action on ProcessingHistory and ErrorLogger will wait until have your input.

**Comment by wangyinz** [comment]
Well, the reason why I suggested moving elog to the Core objects is that we can then simplify the "development guide" with such a rule that the C++ code will always use CoreTimeSeries and CoreSeismogram and the Python code will only use TimeSeries and Seismogram. We can make elog an optional member of the Core objects so that an algorithm can ignore it if not necessary and use the standard exception handling interface. Note that whether the C++ code uses elog or not,, any exceptions it throws will be captured on the Python side so we don't need to implement an additional logic to handle the two different scenarios. 

The reason I don't want a C++ code to deal with TimeSeries and Seismogram is that we want to avoid the potential problems of maintaining two versions of history handling code on the C++ and Python side. But maybe I was thinking it wrong and there is no problems of handing history on the C++ side and transferring it to the Python side. I really need to look at an example to understand how it behaves.

Anyway, if you are hundred percents sure that it is better to put the elog under TimeSeries and Seismogram, I think we should just go for it.

**Comment by pavlis** [comment]
Maybe not the best example, but look in [CNR3CDecon.cc](https://github.com/wangyinz/mspass/blob/master/cxx/src/lib/algorithms/deconvolution/CNR3CDecon.cc) at the function TestSeismogramInput.   It tests for multiple possible errors and posts all of them to elog before returning an error count.   The processor running the test then handles a nonzero error code.   There are other similar examples where the fundamental need is to log any error of severity below what we are calling Invalid.   i.e. an error that doesn't invalidate the result but for which the user might need to be aware of  to track down a later problem or use as a QC filter.   

A very good reason to separate elog and ProcessingHistory is that I do think no C code should handle ProcessingHistory until we find a compelling example that needs to do so.   As I said, elog is a different matter as the above example illustrates.  

In general rules are meant to be broken and I think we should not impose any more of them than are essential.   I am going to make the following changes in the branch I'm working on and push them to github with a pull request where we can finalize this:
1.  elog is going to be a direct member now of Seismogram and TimeSeries instead of being a component of ProcessingHistory.
2. ProcessingHistory::clear is now aliased to clear_history by the pybind11 code.
3. Metadata::clear is now aliased to erase by the pybind11 code. 
4. CoreSeismogram and CoreTimeSeries will be aliased as _CoreSeismogram and _CoreTimeSeries respectively.

Those changes probably won't take long so by the time you read this it may have appeared as a pull request in github.  If I missed something record it here for the record and I'll push a revision before we move the conversation to the pull branch.

**Comment by pavlis** [comment]
I've ran into an anomaly in testing CNR3CDecon that suggests strongly the point made here about never posting elog messages inside C++ code may be a good idea for a different reason.   I was running a test of CNR3CDecon that intentionally caused an error in the TestSeismogramInput method of that processor.   The detail doesn't totally matter, but for the record in gdb I can see this section of code being executed:
```
  if(this->processing_window.start<d.t0() || this->processing_window.end>d.endtime())
  {
    stringstream ss;
    ss<<base_error<<"Data time window mistmatch."<<endl
            <<"Data span relative time range ="<<d.t0()<<" to "<<d.endtime()<<endl
            <<"Processing window range of "<<this->processing_window.start
            << " to " <<this->processing_window.start<<" is not inside data range"<<endl;
    d.elog.log_error("CNR3CDecon",ss.str(),ErrorSeverity::Invalid);
    ++error_count;
  }
```
The caller of this method has a handler for error_count return being nonzero.  This is it:
```
    errcount=TestSeismogramInput(d,-9999,nload);
    if(errcount>0)
    {
      stringstream ss;
      ss<<"CNR3CDecon::loaddata:  "<<errcount<<" errors were detected in this call"
        <<endl<<"Check error log of input Seismogram for detailed error messages"<<endl
        << "Operator does not contain valid data for processing"<<endl;
      throw MsPASSError(ss.str(),ErrorSeverity::Invalid);
    }
```
What is weird is that when I call the loaddata method that generates this error with this line in python:
```
decon.loaddata(d2,False)
```
and look at the contents of the error log I get this:
```
print(d2.elog.size())
0
```
which says the error log is empty.   

I have tested posting an error to the same d2 within python and that creates and entry.  Further, I was worried about the new implementation moving elog out of ProcessingHistory and problems I'd had with the copy constructor so I did this test of data with the elog manually added:
```
dtest=Seismogram(d2)
```
I found dtest had a preserved copy of the log entry I had posted in python.

I am somewhat mystified by this problem.  The only hypothesis I can come up with is that this is a python memory manager issue.  This is the signature for the method being called here:
```
void CNR3CDecon::loaddata(Seismogram& d,const bool nload)
```
so it when d is altered by posting a message to elog the result should appear on the python side.  I should note that in gdb I can verify that the error log has data before the MsPASSError is thrown.  My working hypothesis is that maybe when an error is thrown some hidden copy function is disabled?   I am not sure.   

An alternate hypothesis I need to rule out is that there remains some weird bug in the new implementation with elog an attribute of Seismogram.   I guess I need a more focused test.   I think I will do that by additions to test_tcs to rule out issues on the C++ side.  (Note as stated above copying works fine in the pure python side).

**Comment by wangyinz** [comment]
If you are running the test in the "cleanup_core_C_api" branch, this is probably caused by the same problem that I posted last night here: https://github.com/wangyinz/mspass/pull/94#issuecomment-749393096

**Comment by pavlis** [comment]
I think you may have said something about this earlier, but I just discovered something we absolutely have to fix.   One can innocently try:
```
x=Seismogram(100)
y=Seismogram(100)
z=x+y
# or 
x+=y
```
Neither give what is expected.   I think in both cases you Metadata::operator+ and Metadata::operator+= respectively.   More surprising to me was that even if I did this with CoreTimeSeries where operator+= is at least defined in the C++ code it doesn't work.  The reason, of course, is that those operators are only defined in the pybind11 code for Metadata.   We need to fix this as it could cause mysterious behavior for any user and invalid results.

I will act on this after Christmas, but I seem to also remember you may have addressed part of this problem already. If so, please remind me but I'm assuming I need to just make any operator defined for base classes of Seismogram and TimeSeries has a TimeSeries and Seismogram equivalent.

**Comment by wangyinz** [comment]
We also need to bind these operators in the pybind11 code. I don't think I've fixed this issue before.

**Comment by pavlis** [comment]
Working on the operators and discovered a something we MUST be cautious about in the C++ part of this code base.  The issue is created by the somewhat obnoxious warning error gcc seems to create by default about comparisons of signed and unsigned quantities.    As a result one of us changed this method in BasicTimeSeries  to the following:
```
size_t sample_number(double t)const
        {
            return(round((t-mt0)/mdt));
        };
```
noting the size_t is an alias for `unsigned long int` or `uint64`.   

The problems is there are legitimate cases where you this to return a negative.   Here is the example that bit me in operator+=:
```
i=d.sample_number(this->mt0);
    if(i<0)
    {
      j0=0;
      i0=(-i);
    }
    else
    {
      j0=0;
      i0=i;
    }
```
The test for i is appropriate for a negative because it means the right hand side has a start time before the start time of the lhs of the += operator.   When I tested this with sample_number() returning a size_t when the return resolved to a negative I got the (known to me but obscure to users) unexpected return of a large number.  That happens, of course, because the standard negative ints on every machine these days if converted to unsigned as bit pattern will resolve to large numbers because they have the 64th bit set 1.  

I am going through the seismic components (i.e. TimeSeries and Seismogram and their ancestors) and attempting to remove all potentially improper uses of size_t like that.   

I think it is a bit worse interacting with python because I do not thing unsigned ints are native types in python - they can be defined by they aren't native are they?
